### PR TITLE
wf: implement windows firewall using inet.af/wf.

### PR DESCRIPTION
**Pull request by maisem**
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

**Comment by maisem** [comment]
Need to pull in the latest changes from `inet.af/wf` for the tests/build to pass.
Waiting on inetaf/wf#5, inetaf/wf#6, inetaf/wf#7

File: go.mod [hunk 1]
```diff
@@ -41,7 +41,10 @@ require (
 	inet.af/netaddr v0.0.0-20210222205655-a1ec2b7b8c44
 	inet.af/netstack v0.0.0-20210317161235-a1bf4e56ef22
 	inet.af/peercred v0.0.0-20210302202138-56e694897155
+	inet.af/wf v0.0.0-20210315002834-2cd3664404c3
 	rsc.io/goversion v1.2.0
 )
 
 replace github.com/github/certstore => github.com/cyolosecurity/certstore v0.0.0-20200922073901-ece7f1d353c2
+
+replace inet.af/wf => ../../inet.af/wf
```
Comment (bradfitz) [inline go.mod:50]: delete

File: wf/firewall.go [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+		permittedRoutes: make(map[netaddr.IPPrefix][]*wf.Rule),
+	}
+	if err := f.enable(); err != nil {
+		return nil, err
+	}
+	return f, nil
+}
+
+func (f *Firewall) enable() error {
+	if err := f.permitTailscaleService(15); err != nil {
```
Comment (danderson) [inline wf/firewall.go:135]: These function calls contain magic numbers. Define constants for them?

File: wf/firewall.go [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+	return rules, nil
+}
+
+func (f *Firewall) blockAll(weight uint64) error {
+	_, err := f.addRules("all", weight, nil, wf.ActionBlock, protocolAll, directionBi)
+	return err
+}
+
+func (f *Firewall) permitNDP(weight uint64) error {
+	fieldICMPType := wf.FieldIPLocalPort
```
Comment (danderson) [inline wf/firewall.go:264]: Wow, really? That's a bit of an API wart :/

**Review by danderson — APPROVED** [review]
I assume you've tested that tailscale doesn't fall apart on windows with this change :) The code looks right, but wf is relatively untested, so there's uncertainty there.

We probably want to delay merging this until we've branched off 1.8, it's a scary change to add at the last second.

File: wf/firewall.go (hunk 3 above)
Comment (maisem) [reply wf/firewall.go:267]: Heh, yeah. Added a comment with this link
https://social.msdn.microsoft.com/Forums/azure/en-US/eb2aa3cd-5f1c-4461-af86-61e7d43ccc23/filtering-icmp-by-type-amp-code?forum=wfp

**Comment by maisem** [comment]
> I assume you've tested that tailscale doesn't fall apart on windows with this change :) The code looks right, but wf is relatively untested, so there's uncertainty there.

Yep, I've tested that it doesn't fall apart :) 

> We probably want to delay merging this until we've branched off 1.8, it's a scary change to add at the last second.

I agree, this PR doesn't actually start using the new code path. It's basically dead code for now.

**Pull request merged**